node/stateSync.go

```diff
@@ -0,0 +1,201 @@
+package node
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/nknorg/nkn/v2/chain/db"
+	"github.com/nknorg/nkn/v2/util/log"
+
+	"github.com/nknorg/nkn/v2/config"
+
+	"github.com/nknorg/nkn/v2/chain/trie"
+	"github.com/nknorg/nkn/v2/common"
+)
+
+type stateSync struct {
+	d *Downloader // Downloader instance to access and manage current peerset
+
+	root   common.Uint256 // State root currently being synced
+	sched  *trie.Sync     // State trie sync scheduler defining the tasks
+	hasher func([]byte) []byte
+
+	done chan struct{} // Channel to signal termination completion
+	err  error         // Any error hit during sync (set before completion)
+
+	hashesCh       chan []common.Uint256
+	stateCh        chan [][]byte // Response states channel
+	timeoutReq     chan []common.Uint256
+	failedHashesCh chan []common.Uint256
+	stopChans      []chan struct{}
+
+	peers []*RemoteNode // Set of active peers from which download can proceed
+
+	bytesUncommitted int
+}
+
+// loop is the main event loop of a state trie sync. It is responsible for the
+// assignment of new tasks to peers (including sending it to them) as well as
+// for the processing of inbound data.
+func (s *stateSync) loop() (err error) {
+	defer func() {
+		cerr := s.commit(true)
+		if err == nil {
+			err = cerr
+		}
+		cerr = s.d.database.BatchPut(db.CurrentStateTrie(), s.root.ToArray())
+		if err == nil {
+			err = cerr
+		}
+		if err := s.d.database.BatchCommit(); err != nil {
+			err = cerr
+		}
+	}()
+
+	go s.startWorkerThread()
+
+	go func() {
+		for {
+			log.Info("pending states:", s.sched.Pending())
```

Comment:
Is this for debug only, or we plan to add this log for release version?

Reply:
For debug only